packages/muelu/adapters/stratimikos/Thyra_MueLuRefMaxwellPreconditionerFactory_def.hpp

```diff
@@ -207,11 +213,13 @@ namespace Thyra {
 
         if (paramList.isParameter("M1")) {
           if (paramList.isType<Teuchos::RCP<TpCrsMat> >("M1")) {
```

Comment:
Is it possible that M1 is neither a TpCrsMat or a ThyLinOpBase?

Reply:
In the _decl, we have a specialization for Epetra that checks for Epetra matrix type as well, if that's what you are referring to.

Reply:
@cgcgcg What I meant is that the `if-then` in file `Thyra_MueLuRefMaxwellPreconditionerFactory_def.hpp`  line 214 is of the form
```cpp
         if (paramList.isParameter("M1")) {
￼            if (paramList.isType<Teuchos::RCP<TpCrsMat> >("M1")) {
              Teuchos::TimeMonitor tM_M1(*Teuchos::TimeMonitor::getNewTimer(std::string("ThyraMueLuRefMaxwell::initializePrec wrap M1")));
￼              RCP<TpCrsMat> tM1 = paramList.get<RCP<TpCrsMat> >("M1");
￼              paramList.remove("M1");
￼              RCP<XpCrsMat> xM1 = rcp_dynamic_cast<XpCrsMat>(tM1, true);
￼              paramList.set<RCP<XpCrsMat> >("M1", xM1);
￼            } else if (paramList.isType<Teuchos::RCP<const ThyLinOpBase> >("M1")) {
￼              Teuchos::TimeMonitor tM_M1(*Teuchos::TimeMonitor::getNewTimer(std::string("ThyraMueLuRefMaxwell::initializePrec wrap M1")));
￼              RCP<const ThyLinOpBase> thyM1 = paramList.get<RCP<const ThyLinOpBase> >("M1");
￼              paramList.remove("M1");
￼              RCP<const XpCrsMat> crsM1 = XpThyUtils::toXpetra(thyM1);
              TEUCHOS_TEST_FOR_EXCEPT(Teuchos::is_null(crsM1));
              // MueLu needs a non-const object as input
              RCP<XpCrsMat> crsM1NonConst = Teuchos::rcp_const_cast<XpCrsMat>(crsM1);
              TEUCHOS_TEST_FOR_EXCEPT(Teuchos::is_null(crsM1NonConst));
              // wrap as an Xpetra::Matrix that MueLu can work with
              RCP<XpMat> M1 = rcp(new Xpetra::CrsMatrixWrap<Scalar,LocalOrdinal,GlobalOrdinal,Node>(crsM1NonConst));
              paramList.set<RCP<XpMat> >("M1", M1);
           }
```

Could there a case where M1 is a parameter, but neither of the `if` checks is true?

Reply:
Ah, yes. We should throw then.